ee/modules/600-flant-integration/images/flant-pricing/hooks/controller_metrics.py

```diff
@@ -0,0 +1,200 @@
+        json.dump([], f)
+
+
+@dataclass
+class Controller:
+    name: str
+    namespace: str
+    kind: str
+    module: str
+    cpu: Optional[float] = None
+    memory: Optional[float] = None
+
+
+def main(ctx: hook.Context):
+    # Generate list of Controllers from snapshots
+    controllers = process_controllers(snapshots=ctx.snapshots)
+
+    # Generate metrics from Controllers list
+    metrics = generate_metrics(controllers=controllers)
+
+    # Export metrics
+    ctx.metrics.expire(group=PROMETHEUS_METRIC_GROUP)
+    for metric in metrics:
+        ctx.metrics.collect(metric)
+
+
+def generate_metrics(controllers: List[Controller]) -> List[Dict[str, Any]]:
+    '''Generate metrics from Controllers list'''
+
+    metrics = []
+    for ctrl in controllers:
+        metrics.append({
+            "name": PROMETHEUS_CPU_METRIC_NAME,
+            "group": PROMETHEUS_METRIC_GROUP,
+            "set": ctrl.cpu,
+            "labels": controller_metric_labels(ctrl=ctrl),
+        })
+        metrics.append({
+            "name": PROMETHEUS_MEMORY_METRIC_NAME,
+            "group": PROMETHEUS_METRIC_GROUP,
+            "set": ctrl.memory,
+            "labels": controller_metric_labels(ctrl=ctrl),
+        })
+    return metrics
+
+
+def controller_metric_labels(ctrl: Controller):
+    '''Helper func with labels for resource metrics'''
+
+    return {
+        "controller_name": ctrl.name,
+        "controller_module": ctrl.module,
+        "controller_kind": ctrl.kind,
+    }
+
+
+def process_controllers(snapshots: Dict[str, List[Dict[str, Any]]]) -> List[Controller]:
+    '''Generate list of Controllers from binding context snapshots'''
+
+    return [parse_controller(controller_snapshot=sn) for queue_snapshot in snapshots.values() for sn in queue_snapshot]
+
+
+def parse_controller(controller_snapshot: Dict[str, Any]) -> Controller:
+    '''
+    Generate controller from snapshot and
+    query prometheus for its memory and cpu consumption
+    '''
+
+    filter_result = controller_snapshot["filterResult"]
+    controller = Controller(
+        kind=filter_result["controller_kind"],
+        name=filter_result["controller_name"],
+        namespace=filter_result["controller_namespace"],
+        module=filter_result["controller_module"],
+    )
+    cpu_mock_data, mem_mock_data = None, None
+    if check_for_generate_mock():
+        mock_data = {
+            "group": PROMETHEUS_METRIC_GROUP,
+            "labels": controller_metric_labels(ctrl=controller),
+        }
+        cpu_mock_data = dict(**mock_data, name=PROMETHEUS_CPU_METRIC_NAME)
+        mem_mock_data = dict(**mock_data, name=PROMETHEUS_MEMORY_METRIC_NAME)
+
+    controller.cpu = get_cpu_prometheus(
+        controller=controller,
+        mock_data=cpu_mock_data,
+        mock_data_file=MOCK_DATA_FILE,
+    )
+    controller.memory = get_memory_prometheus(
+        controller=controller,
+        mock_data=mem_mock_data,
+        mock_data_file=MOCK_DATA_FILE,
```

Comment:
If we need mocks for tests, the mocks should live in tests and should not pollute business logic.

I suggest making facade class that abstracts how you get the mertics. It is a single class with single method (get/fetch/whatever). Its instance can be passed as a dependency. The implementation of such an interface with fixture (another class) can be used in tests. It would simplify the code and separate concerns better.